### PR TITLE
Fix BART

### PR DESCRIPTION
**Pull request by epwalsh**
Getting good results now:

```python
# tmp.py
from allennlp.data import Instance, Vocabulary, Batch
from allennlp.data.fields import TextField
from allennlp.data.tokenizers import PretrainedTransformerTokenizer
from allennlp.data.token_indexers import PretrainedTransformerIndexer
from allennlp_models.generation.models import Bart
from allennlp.nn.util import move_to_device

model_name = "facebook/bart-large-cnn"
ARTICLE_TO_SUMMARIZE = """
The tower is 324 metres (1,063 ft) tall, about the same height as an 81-storey building,
and the tallest structure in Paris.
Its base is square, measuring 125 metres (410 ft) on each side.
During its construction, the Eiffel Tower surpassed the Washington Monument to
become the tallest man-made
structure in the world, a title it held for 41 years until the Chrysler Building in
New York City was finished in 1930.
It was the first structure to reach a height of 300 metres.
Due to the addition of a broadcasting aerial at the top of the tower in 1957,
it is now taller than the Chrysler Building by 5.2 metres (17 ft).
Excluding transmitters, the Eiffel Tower is the second tallest free-standing structure in
France after the Millau Viaduct.
""".strip().replace("\n", " ")

tokenizer = PretrainedTransformerTokenizer(model_name)
token_indexers = {"tokens": PretrainedTransformerIndexer(model_name)}
vocab = Vocabulary.empty()
model = Bart(model_name, vocab, indexer=token_indexers["tokens"]).cuda().eval()


def predict(text: str):
    tokens = tokenizer.tokenize(text)
    instance = Instance({"source_tokens": TextField(tokens, token_indexers)})
    instance.index_fields(vocab)
    batch = Batch([instance])
    tensor_dict = move_to_device(batch.as_tensor_dict(), 0)
    result = model(**tensor_dict)
    string = (
        "".join([t.text for t in result["predicted_tokens"][0]])
        .replace("Ġ", " ")
        .replace("<s>", "")
        .replace("</s>", "")
    )
    return string


print(predict(ARTICLE_TO_SUMMARIZE))
```

```
$> python tmp.py
The tower is 324 metres (1,063 ft) tall, about the same height as an 81-storey building.
Its base is square, measuring 125 metres (410 ft) on each side.
```

### CODE REVIEW

**Comment by epwalsh** [comment]
Tests will fail until https://github.com/allenai/allennlp/pull/4872 is merged.

File: allennlp_models/generation/models/bart.py [hunk 1]
```diff
@@ -263,29 +264,28 @@ def forward(
...
     @staticmethod
-    def _decoder_cache_to_dict(decoder_cache):
+    def _decoder_cache_to_dict(decoder_cache: DecoderCacheType) -> Dict[str, torch.Tensor]:
         cache_dict = {}
         for layer_index, layer_cache in enumerate(decoder_cache):
-            for attention_name, attention_cache in layer_cache.items():
-                for tensor_name, cache_value in attention_cache.items():
-                    key = (layer_index, attention_name, tensor_name)
-                    cache_dict[key] = cache_value
+            assert len(layer_cache) == 4
```
Comment (dirkgr) [inline allennlp_models/generation/models/bart.py:270]: Why are there four layers? I assume this is related to the internal huggingface changes?

**Review by dirkgr — APPROVED** [review]
Thanks for tackling this!

I assume we need to re-train the pre-trained model?

File: allennlp_models/generation/models/bart.py (hunk 1 above)
Comment (epwalsh) [reply allennlp_models/generation/models/bart.py:270]: Yea, this is internal to HF. Each layer cache contains the key and value tensors for the self-attention and cross-attention. So 4 tensors per layer. I'll add a comment on that.

**Comment by epwalsh** [comment]
> I assume we need to re-train the pre-trained model?

These changes only affect validation / prediction, so I don't think it's necessary to retrain. What model are you talking about though? Did we actually fine-tune BART on something? It might be worth re-training since the validation metric was probably off.
